src/providers/token-properties-provider.ts

```diff
@@ -103,27 +105,51 @@ export class TokenPropertiesProvider implements ITokenPropertiesProvider {
           tokenValidationResult: TokenValidationResult.UNKN,
         };
       } else {
-        addressesToFetchFeesOnchain.push(address);
+        addressesToFetchFees.push(address);
       }
     }
 
-    if (addressesToFetchFeesOnchain.length > 0) {
+    if (addressesToFetchFees.length > 0) {
       let tokenFeeMap: TokenFeeMap = {};
 
       try {
         tokenFeeMap = await this.tokenFeeFetcher.fetchFees(
-          addressesToFetchFeesOnchain,
+          addressesToFetchFees,
           providerConfig
         );
       } catch (err) {
         log.error(
           { err },
-          `Error fetching fees for tokens ${addressesToFetchFeesOnchain}`
+          `Error fetching fees for tokens ${addressesToFetchFees}`
         );
       }
 
+      if (this.tokenFeeFetcherFallback) {
+        // If a fallback fetcher is provided, we will use it to fetch fees for tokens that were not fetched by the primary fetcher.
+        const addressesToFetchFeesWithFallbackFetcher = addressesToFetchFees.filter(
+          (address) => !tokenFeeMap[address]
+        );
+        if (addressesToFetchFeesWithFallbackFetcher.length > 0) {
+          try {
+            const tokenFeeMapFromFallback = await this.tokenFeeFetcherFallback.fetchFees(
+              addressesToFetchFeesWithFallbackFetcher,
+              providerConfig
+            );
+            tokenFeeMap = {
+              ...tokenFeeMap,
+              ...tokenFeeMapFromFallback,
+            };
+          } catch (err) {
+            log.error(
+              { err },
+              `Error fetching fees for tokens ${addressesToFetchFeesWithFallbackFetcher} using fallback`
+            );
+          }
+        }
+      }
+
```

Comment:
What do you think about creating a `GraphQlTokenFeeFetcherWithFallback` implementation of `ITokenFeeFetcher`, and we put that in `routing-api` instead of here?

That way we could avoid all changes in SOR, we just need to make sure that the `tokenFeeFetcher` can be overwritten in routing-api.

Reply:
that's not a bad idea, let me do that.

Reply:
I like @mikeki 's idea. Then in routing-api, we can create a traffic switch fetcher class (similar to lib/handlers/quote/provider-migration/v3/traffic-switch-on-chain-quote-provider.ts), that can achieve:
1. beginning by shadow calling GQL token fetcher without affecting online path
2. compare the results between GQL token fetcher and direct on-chain token fetcher
3. gradually switching from direct on-chain token fetcher to GQL token fetcher
4. and the existing traffic switch implementation already guarantees that in case GQL token fetcher throws error, it will fallback to use GQL token fetcher.